primitives/src/merkle_tree/field_based_mht/smt/big_lazy_merkle_tree.rs

```diff
@@ -0,0 +1,987 @@
+                }
+
+            },
+            Ok(None) => {
+                return None;
+            },
+            Err(e) => {
+                println!("operational problem encountered: {}", e);
+                return None;
+            },
+        }
+    }
+
+    pub fn check_b_plus_caching_level_down(&mut self, coord: Coord) {
+
+        let left_child_idx = coord.idx * T::MERKLE_ARITY;
+        let left_child_height = 0;
+        let left_coord = Coord { height: left_child_height, idx: left_child_idx };
+
+        let right_child_idx = left_child_idx + 1;
+        let right_child_height = 0;
+        let right_coord = Coord { height: right_child_height, idx: right_child_idx };
+
+        if (!self.state.present_node.contains(&left_coord)) | (!self.state.present_node.contains(&right_coord)) {
+            if self.contains_key_in_cache(coord) {
+                LazyBigMerkleTree::remove_node_from_cache(self, coord);
+            }
+        }
+        return;
+    }
+
+    pub fn check_b_plus_caching(&mut self, coord: Coord) {
+        assert_eq!(T::MERKLE_ARITY, 2, "Arity of the Merkle tree is not 2.");
+
+        if coord.height <= 1 {
+            return;
+        }
+
+        let left_child_idx = coord.idx * T::MERKLE_ARITY;
+        let left_child_height = coord.height - 1;
+        let left_coord = Coord { height: left_child_height, idx: left_child_idx };
+
+        let right_child_idx = left_child_idx + 1;
+        let right_child_height = left_child_height;
+        let right_coord = Coord { height: right_child_height, idx: right_child_idx };
+
+        if left_child_height == 1 {
+            self.check_b_plus_caching_level_down(left_coord);
+            self.check_b_plus_caching_level_down(right_coord);
+            return;
+        }
+    }
+
+    // Returns the hash value associated to the node.
+    // If the node is in the cache, it retrieves from it.
+    // If not, recomputes it.
+    // Only used for nodes of level >= 1 (not for leaves).
+    pub fn node(&mut self, coord: Coord) -> F {
+
+        assert_eq!(T::MERKLE_ARITY,2, "Arity of the Merkle tree is not 2.");
+
+        //let coord = Coord{height, idx};
+        // if the node is an empty node return the hash constant
+        if !self.state.present_node.contains(&coord) {
+            return T::EMPTY_HASH_CST[coord.height];
+        }
+        let res = self.get_from_cache(coord);
+
+        // if not in the cache, recompute it
+        if res == None {
+            /* the node is not in the cache, compute it */
+            let node_hash;
+            if coord.height == 1 {
+                /* get leaves to compute */
+                let left_child_idx = coord.idx * T::MERKLE_ARITY;
+                let left_child = self.get_from_db(left_child_idx);
+                let left_hash: F;
+                if let Some(i) = left_child {
+                    left_hash = i;
+                } else {
+                    left_hash = T::EMPTY_HASH_CST[0];
```

Comment:
I assume this is for safety. But this can't actually happen. If happens there's a bug and we must get notified of this. So let's just call self.get_from_db.unwrap(...) at line 346